Remove unnecessary includes

tautschnig opened the pull request:
Manual removal of includes based on include-what-you-use's output, filtered for includes that should be removed. The goal is to avoid unnecessary build dependencies, reducing the amount of code that needs to be rebuilt during incremental builds.

This is a second iteration of this cleanup; the first round took place in 450845dbc5.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Remove unnecessary includes

Manual removal of includes based on include-what-you-use's output,
filtered for includes that should be removed. The goal is to avoid
unnecessary build dependencies, reducing the amount of code that needs
to be rebuilt during incremental builds.

This is a second iteration of this cleanup; the first round took place in
450845dbc5.

feliperodri reviewed (approved):
Thanks for this clean up!

tautschnig commented:
@chris-ryder Could you please take a look at this one?

tautschnig committed:
Mark includes that include-what-you-use spuriously suggests to remove

This is to make sure that we can run include-what-you-use without being
told to remove includes when doing so would fail the build.

chris-ryder reviewed (approved):
Looks like a painful, but very useful cleanup!

tautschnig committed:
Add validate to symbol_table_baset

This enables use of symbol_table_baset instead of symbol_tablet in many
places.

tautschnig committed:
Use symbol_table_baset instead of symbol_tablet where possible

We almost never create a fresh symbol table, and any time we do so we
should really make sure this is the intended behaviour.

tautschnig committed:
Do not construct a namespace ad hoc when the environment provides one

The empty namespace will lack symbol information that may be needed, and
is just unnecessary anyway.

tautschnig committed:
Remove stray whitespace in comment header

Whitespace cleanup only so as not to break syntax highlighting.

tautschnig committed:
Make format unit test do something

6b88028ba1653 removed the existing test (for good reasons) but didn't
replace it with any other.

tautschnig committed:
Add include-what-you-use as GitHub action

This is to make sure that future pull requests do not introduce
unnecessary includes (or remove ones that are no longer necessary).

TGWDB commented:
@tautschnig  As a member of @diffblue/diffblue-opensource I approved before when it passed checks - is something else needed?

tautschnig committed:
memory-analyzer unit test: include the header file, not cpp

There was a spurious include of the implementation file rather than the
header.

tautschnig commented:
> @tautschnig As a member of @diffblue/diffblue-opensource I approved before when it passed checks - is something else needed?

I've now added a GitHub action to run include-what-you-use, I believe this triggered the additional review request.

thomasspriggs commented on src/solvers/smt2_incremental/smt_response_validation.h:
```diff
@@ -3,11 +3,14 @@
+#include <util/irep.h>
 #include <util/nodiscard.h>
 
-#include <solvers/smt2_incremental/ast/smt_responses.h>
```
`response_or_errort` handles the type it is specialised with by-value like optional does. `response_or_errort` is specialised with `smt_responset` for the return value of `validate_smt_response`. Therefore I would say that it is incorrect to remove this include as it is used.

tautschnig replied:
I have now added `// IWYU pragma: keep` instead.

thomasspriggs commented on src/solvers/smt2_incremental/smt_to_smt2_string.cpp:
```diff
@@ -13,7 +13,6 @@
 
 #include <functional>
 #include <iostream>
-#include <sstream>
```
`std::stringstream` is used in the `smt_to_smt2_string` function (all overloads) and is defined in the `<sstream>`  header. Therefore this include should not be removed, as it is used. I assume this file still compiles without the include due to transitive includes of some type. My understanding is that some of these transitive includes can be in implementations of the STL but not required by the standard.

tautschnig replied:
I have now added `// IWYU pragma: keep` instead.

thomasspriggs commented on src/solvers/smt2_incremental/type_size_mapping.h:
```diff
@@ -6,13 +6,12 @@
-#include <util/expr.h>
+#include <util/expr.h> // IWYU pragma: keep
 
-#include <solvers/smt2_incremental/ast/smt_terms.h>
```
`smt_termt`is defined in `smt_terms.h` and is used in the definition of `type_size_mapt` in this file. Therefore this include should be kept.

tautschnig replied:
I have now added `// IWYU pragma: keep` instead.

thomasspriggs commented on unit/solvers/smt2_incremental/smt_to_smt2_string.cpp:
```diff
@@ -1,9 +1,8 @@
+#include <util/mp_arith.h> // IWYU pragma: keep
 
 #include <solvers/smt2_incremental/ast/smt_commands.h>
-#include <solvers/smt2_incremental/ast/smt_logics.h>
```
The logics defined in this header are used as part of the "Test smt_set_logic_commandt to string conversion" test in this file. Removing the include still compiles as it depends on a transitive include via `smt_commends.h`.

tautschnig replied:
I have now added `// IWYU pragma: keep` instead.

thomasspriggs commented on src/goto-symex/symex_main.cpp:
```diff
@@ -361,28 +360,27 @@ void goto_symext::symex_with_state(
     symex_threaded_step(state, get_goto_function);
     if(should_pause_symex)
-      return;
+      return state.symbol_table;
```
I am not convinced this commit is actually using any copy-elision. Are the performance characteristics still the same?

tautschnig replied:
Actually I should not have included this in here, it was just drive-by cleanup. Commit now removed and instead PR-ed as #7275. Perhaps you can elaborate on that new PR?

martin-cs reviewed (approved):
All areas I am maintainer for look good.
